packages/multichain-account-service/src/MultichainAccountWallet.ts

```diff
@@ -438,44 +436,43 @@ export class MultichainAccountWallet<
                * coverage.
                */
               if (
-                !qCtx.stopped &&
-                !qCtx.running &&
-                qCtx.groupIndex < maxGroupIndex
+                !nextContext.stopped &&
+                !nextContext.running &&
+                nextContext.groupIndex < maxGroupIndex
               ) {
-                qCtx.running = schedule(q, maxGroupIndex);
+                nextContext.running = schedule(nextContext, maxGroupIndex);
               }
             }
           }
-        } catch (err) {
-          providerCtx.stopped = true;
-          console.error(err);
+        } catch (error) {
+          context.stopped = true;
+          console.error(
+            `Discovery FAILED for "${context.provider.getName()}" provider:`,
+            error,
+          );
         } finally {
-          providerCtx.running = undefined;
-          if (!providerCtx.stopped) {
-            const target = Math.max(maxGroupIndex, providerCtx.groupIndex);
-            providerCtx.running = schedule(p, target);
+          context.running = undefined;
+          if (!context.stopped) {
+            const highestGroupIndex = Math.max(
+              maxGroupIndex,
+              context.groupIndex,
+            );
+            context.running = schedule(context, highestGroupIndex);
           }
         }
       })();
     };
 
-    const currentGroupIndex = this.getNextGroupIndex();
     // Start discovery for all providers.
-    for (const p of providers) {
-      const pCtx = providerContexts.get(p) as {
-        stopped: boolean;
-        running?: Promise<void>;
-        groupIndex: number;
-        count: number;
-      };
-      pCtx.running = schedule(p, currentGroupIndex);
+    for (const context of providerContexts) {
+      context.running = schedule(context, maxGroupIndex);
     }
 
     let racers: Promise<void>[] = [];
     do {
       racers = [...providerContexts.values()]
-        .map((c) => c.running)
-        .filter(Boolean) as Promise<void>[];
+        .filter((ctx) => ctx.running !== undefined)
+        .map((ctx) => ctx.running) as Promise<void>[];
```

Comment:
I thought doing the `filter` before would reduce the number of iterations, instead of `map` (N) + `filter` (N -> M) we do `filter (N -> M)` + `map` (M).

Obviously, this is extremely nitpick given the number of providers we have 🙈